lib/ruby_lsp/executor.rb

```diff
@@ -202,11 +204,12 @@ def hover(uri, position)
       end
 
       # Instantiate all listeners
-      base_listener = Requests::Hover.new(document.uri, @message_queue)
-      listeners = Requests::Hover.listeners.map { |l| l.new(document.uri, @message_queue) }
+      emitter = EventEmitter.new
+      base_listener = Requests::Hover.new(emitter, @message_queue)
+      listeners = Requests::Hover.listeners.map { |l| l.new(emitter, @message_queue) }
```

Comment:
I wonder if it's possible to encapsulate these `<Listener>.new`, `<Listener>.listeners.map` boilerplate into something like: `<Listener>.run`?

In this example, I imagine the final result would be:

```rb
emitter = EventEmitter.new
# returns a response
Requests::Hover.run(emitter, @message_queue) do |emitter|
  emitter.emit_for_target(target)
end

# listener.rb
class Listener
  # maybe with uri?
  def self.run(emitter, queue)
    base_listener = self.new(emitter, queue)
    listeners = listeners.map { |l| l.new(emitter, queue) }
    yield(emitter)
    listeners.each { |ext| base_listener.merge_response!(ext) }
    base_listener.response
  end
end
```

Reply:
Which request could use both listener classes at the same time. For individual requests, the pattern I proposed should still help?

Reply:
We can follow up with another refactor. One possibility would be to include our own listener as the first entry in the list and then merge all responses into the first. Something like this
```ruby
# document_symbol
class DocumentSymbol < Listener
  add_listener(self)
end

# executor
emitter = EventEmitter.new
document_symbol_listeners = DocumentSymbol.listeners.each { |l| l.new(emitter, @message_queue) }
document_link_listeners = DocumentLink.listeners.each { |l| l.new(emitter, @message_queue) }
emitter.visit(document.tree)
# then use some loop to merge responses into the first entry in the listeners array
```
What do you think?

Reply:
That's nice. I'll create an issue to make this refactor after this one.